R/yoda_test_linter.R

```diff
@@ -0,0 +1,45 @@
+#' Block obvious "yoda tests"
+#'
+#' Yoda tests test `(expected, actual)` instead of the more common `(actual, expected)`.
+#' This is not always possible to detect statically; this linter focuses on
+#'   the simple case of testing an expression against a literal value, e.g.
+#'   `(1L, foo(x))` should be `(foo(x), 1L)`.
+#'
+#' @evalRd rd_tags("yoda_test_linter")
+#' @seealso
+#'   [linters] for a complete list of linters available in lintr.
+#'   <https://en.wikipedia.org/wiki/Yoda_conditions>
+#' @export
+yoda_test_linter <- function() {
+  Linter(function(source_file) {
+    if (length(source_file$parsed_content) == 0L) {
+      return(list())
+    }
+
+    xml <- source_file$xml_parsed_content
+
+    # catch the following types of literal in the first argument:
+    #   (1) numeric literal (e.g. TRUE, 1L, 1.0, NA) [NUM_CONST]
+    #   (2) string literal (e.g. 'str' or "str")     [STR_CONST]
+    #   (3) arithmetic literal (e.g. 1+1 or 0+1i)    [OP-PLUS or OP-MINUS...]
+    xpath <- "//expr[
+      expr[SYMBOL_FUNCTION_CALL[text() = 'expect_equal' or text() = 'expect_identical' or text() = 'expect_setequal']]
+      and expr[2][NUM_CONST or STR_CONST or ((OP-PLUS or OP-MINUS) and count(expr[NUM_CONST]) = 2)]
+    ]"
```

Comment:
Safeguard against false positives via `and expr[3][not(...)]`?
Otherwise this will always lint things like `expect_equal(1.0, 1.0 + 0i)`.

Alternatively, I'd be fine if the message in the latter case would be "do not compare literals in testthat"?

Reply:
>  expect_equal(1.0, 1.0 + 0i)

seems like a bad test besides for R itself :)

we did encounter a few of these being used as placeholder tests but (1) `expect_true(TRUE)` is better and (2) similar to how we lint for `debug()` & friends, I think it's fine to lint placeholder code that shouldn't really make it "to prod"

Reply:
I agree this should be linted, but the message will be very misleading IMO.

Reply:
That's true... still let's table it for now, by making a note to check if this actually comes up in practice before release. If so, we can customize the error message.